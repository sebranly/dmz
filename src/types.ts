// TODO: remove
export enum Template {
  Typescript = 'typescript',
  Javascript = 'javascript'
}

export type TemplateLanguage = {
  language: Template;
};
