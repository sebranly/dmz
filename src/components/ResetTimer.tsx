import classnames from 'classnames';
import * as React from 'react';
import { APITimer, TimerFrequency, TimeUnit } from '../types';
import { getDailyTime, getNextTime, getWeeklyTime } from '../utils';
import { convertSecondsToTimeValue } from '../utils/convert';
import {
  displayWithTwoDigits,
  getStatusColor,
  getStatusVerb,
  getTimeUnitAbbreviation,
  pluralize,
  titleize
} from '../utils/display';
import { getTimerClasses } from '../utils/tailwind';

export interface ResetTimerProps {
  className?: string;
  currentTimestamp: number;
  timer: APITimer;
}

const ResetTimer: React.FC<ResetTimerProps> = (props) => {
  const { className, currentTimestamp, timer } = props;
  const { name, time, status, frequency } = timer;

  const nextTime = getNextTime(currentTimestamp, time, frequency);
  const remainingSeconds = nextTime - currentTimestamp;

  const isDaily = frequency === TimerFrequency.Daily;
  const subtitle = `They ${getStatusVerb(status)} in`;
  const color = getStatusColor(status);

  const classnamesColor = `text-${color}-500`;
  const classnamesSubtitle = 'font-bold my-1 text-lg';
  const classnamesTitle = classnames(classnamesSubtitle, classnamesColor);
  const classnamesTime = classnames('font-bold pr-1', classnamesColor);

  const {
    [TimeUnit.Day]: days,
    [TimeUnit.Hour]: hours,
    [TimeUnit.Minute]: minutes,
    [TimeUnit.Second]: seconds
  } = convertSecondsToTimeValue(remainingSeconds);

  const isFixedDays = days === 0;
  const isFixedHours = isFixedDays && hours === 0;
  const isFixedMinutes = isFixedHours && minutes === 0;
  const isFixedSeconds = isFixedMinutes && seconds === 0;
  const isFixedReset = isDaily && days === 1 && hours === 0 && minutes === 0 && seconds === 0;

  const resetTimeString = isDaily ? getDailyTime(nextTime) : getWeeklyTime(nextTime);

  const classnamesComponent = getTimerClasses(color, className);
  const commonItemsObj = [
    { value: hours, label: TimeUnit.Hour },
    { value: minutes, label: TimeUnit.Minute },
    { value: seconds, label: TimeUnit.Second }
  ];

  const itemsObj = isDaily ? commonItemsObj : [{ value: days, label: TimeUnit.Day }, ...commonItemsObj];

  const items = itemsObj.map((element: { value: number; label: TimeUnit }) => {
    const { value, label: l } = element;
    const isFixed =
      isFixedReset ||
      (l === TimeUnit.Day && isFixedDays) ||
      (l === TimeUnit.Hour && isFixedHours) ||
      (l === TimeUnit.Minute && isFixedMinutes) ||
      (l === TimeUnit.Second && isFixedSeconds);

    const classnamesValue = classnames('font-bold', {
      [classnamesColor]: isFixed,
      'text-4xl': !isDaily,
      'text-5xl': isDaily
    });

    return (
      <li className="inline-block mx-2.5" key={l}>
        <span className={classnamesValue}>{displayWithTwoDigits(value)}</span>
        <div className="text-sm">{isDaily ? pluralize(l, value) : getTimeUnitAbbreviation(l)}</div>
      </li>
    );
  });

  return (
    <div className={classnamesComponent}>
      <div className={classnamesTitle}>{name}</div>
      <div className={classnamesSubtitle}>{subtitle}</div>
      <ul className="timer-card flex justify-center">{items}</ul>
      <div className="text-sm">
        <div className="flex text-left pl-2.5">
          <div className="grow">{`${titleize(frequency)} ${titleize(status)}:`}</div>
          <div className={classnamesTime}>{resetTimeString}</div>
        </div>
      </div>
    </div>
  );
};

export { ResetTimer };
