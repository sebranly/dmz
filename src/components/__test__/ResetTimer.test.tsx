import * as React from 'react';
import { render } from '@testing-library/react';
import { ResetTimer } from '../ResetTimer';
import { APITime, TimeFrequency, TimeType } from '../../types';

test('ResetTimer', () => {
  const time: APITime = {
    type: TimeType.Reset,
    title: 'Daily Challenges',
    frequency: TimeFrequency.Daily,
    data: [{ time: 1678078800 }]
  };

  const createComponent = (currentTimestamp: number, time: APITime) => (
    <ResetTimer currentTimestamp={currentTimestamp} time={time} />
  );

  const { container: containerDaily } = render(createComponent(1678078801, time));
  expect(containerDaily.childNodes).toMatchSnapshot();

  time.title = 'Weekly Challenges';
  time.frequency = TimeFrequency.Weekly;

  const { container: containerWeekly } = render(createComponent(1678078801, time));
  expect(containerWeekly.childNodes).toMatchSnapshot();
});
