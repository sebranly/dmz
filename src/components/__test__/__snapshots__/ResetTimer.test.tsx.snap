// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetTimer 1`] = `
NodeList [
  <div
    class="bg-neutral-800 border-solid border-2 rounded-lg border-red-500 m-2.5 p-2.5 relative w-72 sm:w-80"
  >
    <div
      class="font-bold my-1 text-lg text-red-500"
    >
      Daily Challenges
    </div>
    <ul
      class="timer-card flex justify-center"
    >
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-5xl"
        >
          23
        </span>
        <div
          class="text-sm"
        >
          hours
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-5xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          minutes
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-5xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          seconds
        </div>
      </li>
    </ul>
  </div>,
]
`;

exports[`ResetTimer 2`] = `
NodeList [
  <div
    class="bg-neutral-800 border-solid border-2 rounded-lg border-red-500 m-2.5 p-2.5 relative w-72 sm:w-80"
  >
    <div
      class="font-bold my-1 text-lg text-red-500"
    >
      Weekly Challenges
    </div>
    <ul
      class="timer-card flex justify-center"
    >
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          06
        </span>
        <div
          class="text-sm"
        >
          days
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          23
        </span>
        <div
          class="text-sm"
        >
          hrs
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          min
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          sec
        </div>
      </li>
    </ul>
  </div>,
]
`;
